### PR TITLE
MAINT: Fix/parametrize Rotation tests.

### PR DESCRIPTION
**Pull request by Carreau**
The test were slightly incorrect in the sens that they were testing that at least one of the combination of angle/sequence/intrinsic-extrinsic angle where emitting a Gimbal Lock warning, instead of each of them.

The warning is now checked _only_ and on _all_ on the line that are supposed to emit such a warning.

And while I was at it I parametrized the tests, so each appear as a separate intem when running with pytest, and extended to all the parametrized test in case it ever have an issue somewhere.

In all of the `angles = np.array(...)` I have the impression that each one should emit a warning and that angle should likely also be a parameter, but it was starting to be a lot (maybe open an issue to parametrize angle?)

### CODE REVIEW

File: scipy/spatial/transform/tests/test_rotation.py [hunk 1]
```diff
@@ -676,101 +671,61 @@ def test_stats(error, mean_max, rms_max):
...
-        [35, 90, 25],
-        [25, -90, 15]])
-
-    with pytest.warns(UserWarning, match="Gimbal lock"):
-        for seq_tuple in permutations('xyz'):
-            # Extrinsic rotations
-            seq = ''.join(seq_tuple)
-            rotation = Rotation.from_euler(seq, angles, degrees=True)
-            mat_expected = rotation.as_matrix()
+    angles = np.array([[45, 90, 35], [35, -90, 20], [35, 90, 25], [25, -90, 15]])
```
Comment (Carreau) [inline scipy/spatial/transform/tests/test_rotation.py:691]:
```suggestion
    angles = np.array([
        [45, 90, 35],
        [35, -90, 20],
        [35, 90, 25],
        [25, -90, 15]])
```

Autoformatter got a bit greedy.

File: scipy/spatial/transform/tests/test_rotation.py [hunk 2]
```diff
@@ -676,101 +671,61 @@ def test_stats(error, mean_max, rms_max):
...
-            rotation = Rotation.from_euler(seq, angles, degrees=True)
-            mat_expected = rotation.as_matrix()
-
-            angle_estimates = rotation.as_euler(seq, degrees=True)
-            mat_estimated = Rotation.from_euler(
-                seq, angle_estimates, degrees=True
-                ).as_matrix()
-
-            assert_array_almost_equal(mat_expected, mat_estimated)
+    angles = np.array([[15, 0, 60], [35, 0, 75], [60, 180, 35], [15, -180, 25]])
```
Comment (Carreau) [inline scipy/spatial/transform/tests/test_rotation.py:711]:
```suggestion
    angles = np.array([
        [15, 0, 60],
        [35, 0, 75],
        [60, 180, 35],
        [15, -180, 25]])
```

Autoformatter a bit overzealous as well..

**Comment by Carreau** [comment]
Another possible refactor is to reify `seq_tuple` and join it into a string/uppsercaseify. Right now the test looks like
```
... asymmetric_axes[False-seq_tuple0] PASSED
```
with a cleanup they could look like
```
... asymmetric_axes['xzy'] PASSED
... asymmetric_axes['ZYX'] PASSED
```

**Comment by Carreau** [comment]
Failure seem unrelated, can someone restart the failing item ?

**Comment by Carreau** [comment]
Gentle poke on this one.

File: scipy/spatial/transform/tests/test_rotation.py [hunk 3]
```diff
@@ -676,29 +671,21 @@ def test_stats(error, mean_max, rms_max):
...
-        angles_quat = rotation.as_euler(seq)
-        angles_mat = rotation._as_euler_from_matrix(seq)
-        assert_allclose(angles, angles_quat, atol=0, rtol=1e-13)
-        assert_allclose(angles, angles_mat, atol=0, rtol=1e-9)
-        test_stats(angles_quat - angles, 1e-16, 1e-14)
-        test_stats(angles_mat - angles, 1e-15, 1e-13)
-
-        # Intrinsic rotations
+    seq = "".join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
+    if intrinsic:
```
Comment (h-vetinari) [inline scipy/spatial/transform/tests/test_rotation.py:675]: Please preserve the comments at least as they are, where sensible (though feel free to improve the description of why the 0,1,0 pattern for example).
```suggestion
    # Extrinsic rotations
    seq = "".join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
    if intrinsic:
```

File: scipy/spatial/transform/tests/test_rotation.py [hunk 4]
```diff
@@ -707,34 +694,22 @@ def test_as_euler_degenerate_asymmetric_axes():
...
-                seq, angle_estimates, degrees=True
-                ).as_matrix()
-
-            assert_array_almost_equal(mat_expected, mat_estimated)
-
-            # Intrinsic rotations
-            seq = seq.upper()
-            rotation = Rotation.from_euler(seq, angles, degrees=True)
-            mat_expected = rotation.as_matrix()
+    seq = "".join(seq_tuple)
```
Comment (h-vetinari) [inline scipy/spatial/transform/tests/test_rotation.py:697]: Idem
```suggestion
    # Extrinsic rotations
    seq = "".join(seq_tuple)
```

File: scipy/spatial/transform/tests/test_rotation.py [hunk 5]
```diff
@@ -707,34 +694,22 @@ def test_as_euler_degenerate_asymmetric_axes():
...
+        seq = seq.upper()
+    rotation = Rotation.from_euler(seq, angles, degrees=True)
+    mat_expected = rotation.as_matrix()
 
-            angle_estimates = rotation.as_euler(seq, degrees=True)
-            mat_estimated = Rotation.from_euler(
-                seq, angle_estimates, degrees=True
-                ).as_matrix()
+    with pytest.warns(UserWarning, match="Gimbal lock"):
+        angle_estimates = rotation.as_euler(seq, degrees=True)
```
Comment (h-vetinari) [inline scipy/spatial/transform/tests/test_rotation.py:704]: This is a nice improvement, to restrict `pytest.warns` to the one place it should trigger. 👍

File: scipy/spatial/transform/tests/test_rotation.py [hunk 6]
```diff
@@ -815,35 +771,26 @@ def test_as_euler_degenerate_compare_algorithms():
 
     idx = angles[:, 1] == 0  # find problematic angles indices
 
+    # Extrinsic rotations
+    seq = "".join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
+    if intrinsic:
+        seq = seq.upper()
+    rot = Rotation.from_euler(seq, angles, degrees=True)
```
Comment (h-vetinari) [inline scipy/spatial/transform/tests/test_rotation.py:778]: I don't mind pulling out the processing, but the comments could be clearer IMO.

```suggestion
    # Extrinsic rotations
    seq = "".join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
    if intrinsic:
        # Intrinsic rotations
        seq = seq.upper()

    rot = Rotation.from_euler(seq, angles, degrees=True)
```

**Review by h-vetinari — COMMENTED** [review]
This basically LGTM, it's a very nice clean-up, thanks! Just some small nits around keeping the comments unambiguous and clear. I guess for people familiar with rotations it's clear that the case of the letters switches on extrinsic/intrinsic, but it doesn't hurt to keep that context in the test for easier digestibility.

**Comment by Carreau** [comment]
> Just some small nits around keeping the comments unambiguous and clear. I guess for people familiar with rotations it's clear that the case of the letters switches on extrinsic/intrinsic, but it doesn't hurt to keep that context in the test for easier digestibility.

Yeah, I'm not too familiar with rotation either. My understanding from the docs is that Upper/lower is intrinsic/Extrinsic (ie global, of related to the objects), and axes in the form  abc/aba are either symmetric or asymmetric. I've updated the comments.

**Comment by Carreau** [comment]
Gentle check-in here, as this is only a refactor for further work.

File: scipy/spatial/transform/tests/test_rotation.py [hunk 7]
```diff
@@ -815,35 +783,30 @@ def test_as_euler_degenerate_compare_algorithms():
 
     idx = angles[:, 1] == 0  # find problematic angles indices
 
+    # Rotation of the form A/B/A are rotation around symmetric
+    # Axes
+    seq = "".join([seq_tuple[0], seq_tuple[1], seq_tuple[0]])
+    if intrinsic:
+        # Extrinsinc rotation (wrt to global world at lower case
+        # Intrinsic (WRT the object itself) lower case.
```
Comment (h-vetinari) [inline scipy/spatial/transform/tests/test_rotation.py:791]:
```suggestion
        # Extrinsinc rotation (wrt to global world at lower case
        # Intrinsic (WRT the object itself) lower case.
```

This has an unclosed bracket and describes both cases as lower-case, which isn't true. The same string appears at least 3 times, with the same issue.

File: scipy/spatial/transform/tests/test_rotation.py [hunk 8]
```diff
@@ -815,35 +783,30 @@ def test_as_euler_degenerate_compare_algorithms():
 
     idx = angles[:, 1] == 0  # find problematic angles indices
 
+    # Rotation of the form A/B/A are rotation around symmetric
+    # Axes
```
Comment (h-vetinari) [inline scipy/spatial/transform/tests/test_rotation.py:787]: This reads like an oddly placed linebreak in a sentence. If it's not, please reformulate or add a period/semicolon/whatever.
```suggestion
    # Rotation of the form A/B/A are rotation around symmetric;
    # Axes
```

**Review by h-vetinari — COMMENTED** [review]
Some details that are off

**Comment by h-vetinari** [comment]
Would appreciate if someone else could double-check here from either the spatial owners or other regular reviewers. Once the above nits are fixed, I'll merge this in a few days, if no other comments/reviews appear.

File: scipy/spatial/transform/tests/test_rotation.py (hunk 7 above)
Comment (Carreau) [reply scipy/spatial/transform/tests/test_rotation.py]: Both should be fixed.

File: scipy/spatial/transform/tests/test_rotation.py (hunk 8 above)
Comment (Carreau) [reply scipy/spatial/transform/tests/test_rotation.py]: Should be fixed in the 3 occurences.

**Review by h-vetinari — APPROVED** [review]
If anyone else wants to have a look please do! Otherwise, planning to merge in ~48h.

And thanks for the clean-up @Carreau!

**Comment by Carreau** [comment]
Do you want me to send an email to scipy-dev ?

**Comment by h-vetinari** [comment]
> Do you want me to send an email to scipy-dev ? (to ask for more reviews)

You're free to do so, but I don't think it's that much of a critical issue. At least, I'm happy to merge it as is.

**Review by tylerjereddy — APPROVED** [review]
Core dev approval and test refactors LGTM as well

**Comment by tylerjereddy** [comment]
Thanks!

**Comment by Carreau** [comment]
Thanks !

**Pull request merged**